File: python_files/dsa/doubly_linked_list.py
class Node:
    def __init__(self, data: int) -> None:
        self.data: int = data
        self.prev: Node | None = None
        self.next: Node | None = None


class DoublyLinkedList:
    def __init__(self) -> None:
        self.size: int = 0
        self.head: Node | None = None

    def prepend(self, data: int) -> None:
        new_node: Node = Node(data)

        if self.head:
            self.head.prev = new_node
        new_node.next = self.head

        self.head = new_node
        self.size += 1

    def append(self, data: int) -> None:
        new_node: Node = Node(data)

        if not self.head:
            self.head = new_node
            self.size += 1
            return

        current: Node | None = self.head
        while current.next:
            current = current.next

        new_node.prev = current
        current.next = new_node
        self.size += 1

    def remove_at(self, index: int) -> None:
        assert self.head
        assert self.size > 0
        assert 0 <= index < self.size

        current: Node | None = self.head
        count: int = 0

        while current:
            if count == index:
                if current.prev:
                    current.prev.next = current.next
                if current.next:
                    current.next.prev = current.prev

                if current == self.head:
                    self.head = current.next

                self.size -= 1
                return

            current = current.next
            count += 1

    def remove(self, data: int) -> None:
        assert self.head
        assert self.size > 0

        current: Node | None = self.head

        while current:
            if current.data == data:
                if current.prev:
                    current.prev.next = current.next
                if current.next:
                    current.next.prev = current.prev

                if current == self.head:
                    self.head = current.next

                self.size -= 1
                return

            current = current.next

        print("data not found")

    def reverse(self) -> None:
        assert self.head
        assert self.size > 0

        current: Node | None = self.head
        prev: Node | None = None
        next: Node | None = None

        while current:
            next = current.next

            current.prev = next
            current.next = prev

            prev = current
            current = next

        self.head = prev

    def print(self) -> None:
        assert self.head
        assert self.size > 0

        print("Next: none -> ", end="")
        current: Node | None = self.head

        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("none")

        print("Prev: none <- ", end="")
        # find last node
        current = self.head
        while current.next:
            current = current.next

        while current:
            print(current.data, end=" <- ")
            current = current.prev
        print("none")


def main() -> None:
    arr: DoublyLinkedList = DoublyLinkedList()
    print()
    print("Append: 1...9")
    for i in range(1, 10):
        arr.append(i)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Prepend: 0")
    arr.prepend(0)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Remove at: 4")
    arr.remove_at(4)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Remove: 8")
    arr.remove(8)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Reverse:")
    arr.reverse()
    arr.print()
    print(f"size: {arr.size}")


if __name__ == "__main__":
    main()

File: c_files/dsa/doubly_linked_list/doubly_linked_list.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "doubly_linked_list.h"

Node *create_node(const int data)
{
    Node *node = malloc(sizeof(Node));
    assert(node != NULL);

    node->data = data;
    node->prev = NULL;
    node->next = NULL;

    return node;
}

void dll_init(DoublyLinkedList *list)
{
    assert(list);

    list->head = NULL;
    list->size = 0;
}

void dll_prepend(DoublyLinkedList *list, const int data)
{
    assert(list);

    Node *new_node = create_node(data);

    new_node->next = list->head;
    if (list->head) {
        list->head->prev = new_node;
    }

    list->head = new_node;
    list->size++;
}

void dll_append(DoublyLinkedList *list, const int data)
{
    assert(list);

    Node *new_node = create_node(data);

    if (!list->head) {
        list->head = new_node;
        list->size++;
        return;
    }

    Node *current = list->head;

    while (current->next) {
        current = current->next;
    }

    new_node->prev = current;
    current->next  = new_node;
    list->size++;
}

void dll_remove_at(DoublyLinkedList *list, const int index)
{
    assert(list && list->head);
    assert(list->size > 0);
    assert(index >= 0 && index < list->size);

    Node *current = list->head;
    int   count   = 0;

    while (current) {
        if (count == index) {
            if (current->prev) {
                current->prev->next = current->next;
            }
            if (current->next) {
                current->next->prev = current->prev;
            }

            if (current == list->head) {
                list->head = current->next;
            }

            free(current);
            list->size--;
            return;
        }

        current = current->next;
        count++;
    }
}

void dll_remove(DoublyLinkedList *list, const int data)
{
    assert(list && list->head);
    assert(list->size > 0);

    Node *current = list->head;

    while (current) {
        if (current->data == data) {
            if (current->prev) {
                current->prev->next = current->next;
            }
            if (current->next) {
                current->next->prev = current->prev;
            }

            if (current == list->head) {
                list->head = current->next;
            }

            free(current);
            list->size--;
            return;
        }

        current = current->next;
    }

    printf("data not found\n");
}

void dll_reverse(DoublyLinkedList *list)
{
    assert(list && list->head);
    assert(list->size > 0);

    Node *current = list->head;
    Node *prev    = NULL;
    Node *temp    = NULL;

    while (current) {
        temp = current->next;

        current->prev = temp;
        current->next = prev;

        prev    = current;
        current = temp;
    }

    list->head = prev;
}

void dll_print(DoublyLinkedList *list)
{
    assert(list && list->head);

    if (list->size == 0) {
        printf("empty list\n");
        return;
    }

    printf("Next: null -> ");

    Node *current = list->head;
    while (current) {
        printf("%d -> ", current->data);

        current = current->next;
    }

    printf("null\n");

    //
    // find last node
    current = list->head;
    while (current->next) {
        current = current->next;
    }

    printf("Prev: null <- ");

    while (current) {
        printf("%d <- ", current->data);

        current = current->prev;
    }

    printf("null\n");
}

void dll_free(DoublyLinkedList *list)
{
    assert(list && list->head);
    assert(list->size > 0);

    Node *current = list->head;
    while (current) {
        Node *temp = current;
        current    = current->next;
        free(temp);
    }

    list->head = NULL;
    list->size = 0;
}
